sklearn/feature_extraction/_dict_vectorizer.py

```diff
@@ -376,6 +379,21 @@ def get_feature_names(self):
         """
         return self.feature_names_
 
+    def get_feature_names_out(self, input_features=None):
+        """Get output feature names for transformation.
+
+        Parameters
+        ----------
+        input_features : array-like of str or None, default=None
+            Not used, present here for API consistency by convention.
+
+        Returns
+        -------
+        output_feature_names : list of str
+            Feature names for transformer output.
+        """
+        return self.feature_names_
```

Comment:
This one is concerning because the feature names can be integers:

```py
from sklearn.feature_extraction import DictVectorizer
X = [{'foo': 1, 'bar': 2}, {'foo': 3, 'baz': 1}]
X = [{1:2, 3:4}, {2: 4}]
dv = DictVectorizer(sparse=False).fit(X)
dv.get_feature_names()
# [1, 2, 3]
```

Reply:
I would deprecate this behaviour but this should be done in another PR.

Reply:
You mean deprecate support for non-string keys in dicts passed to `DictVectorizer`?

Alternatively we could call `str()` on the non-str keys to build the input and output feature names for this transformer. But I am not sure how do deal with backward compat warnings...

Reply:
Yes. I think that the discussion was also raised in the PR regarding the column names consistency: https://github.com/scikit-learn/scikit-learn/pull/18010#discussion_r493831843

Reply:
It's one think to say that feature names returned by estimators are always expected to be `str` and it's another to make `DictVectorizer` reject `{2: 1, 4: 2}, {3: 3, 5: 1}` as valid input data in my opinion.

Reply:
> Alternatively we could call str() on the non-str keys to build the input and output feature names for this transformer. But I am not sure how do deal with backward compat warnings...

Since this PR adds a introduces a new method `get_feature_names_out`, we can introduce this new `str()` behavior without needing to worry about backward compatibility.